analyses/mwi/dryspells_trigger/exploration/ecmwf_realtime_test.md

```diff
@@ -0,0 +1,256 @@
+import os
+import sys
+
+import pandas as pd
+import numpy as np
+
+import xarray as xr
+import fnmatch
+import geopandas as gpd
+
+from rasterio.enums import Resampling
+from matplotlib.colors import ListedColormap
+
+path_mod = f"{Path(os.path.dirname(os.path.abspath(''))).parents[2]}/"
+sys.path.append(path_mod)
+from src.indicators.drought.config import Config
+from src.indicators.drought.ecmwf_seasonal import processing
+from src.utils_general.raster_manipulation import compute_raster_statistics
+reload(processing);
+```
+
+#### Set config values
+
+```python
+country_iso3="mwi"
+config=Config()
+parameters = config.parameters(country_iso3)
+
+country_data_raw_dir = os.path.join(config.DATA_DIR,config.PUBLIC_DIR, config.RAW_DIR,country_iso3)
+adm1_bound_path=os.path.join(country_data_raw_dir,config.SHAPEFILE_DIR,parameters["path_admin1_shp"])
+ecmwf_realtime_dir = Path(config.DATA_DIR) / config.PRIVATE_DIR / config.RAW_DIR / "glb" / "ecmwf"
+```
+
+```python
+gdf_adm1=gpd.read_file(adm1_bound_path)
+```
+
+We select all files that contain the seasonal forecast with monthly mean, i.e. that Have T4L in their filename
+
+```python
+filepath_list = [
+        filename
+        for filename in ecmwf_realtime_dir.iterdir()
+        #T4L indicates the seasonal forecast with the monthly mean
+        #see here for a bit more explanation https://confluence.ecmwf.int/pages/viewpage.action?pageId=111155348
+        #without the 1 you also get .idx files which we don't want but not sure if this is the best method to select
+        if fnmatch.fnmatch(filename, '*T4L*1') 
+    ]
+```
+
+```python
+filepath_list
+```
+
+We first load one file without any concatenation and processing to understand the fileformat.  
+
+The files are very cryptic but they contain a dataType which can be "em" or "fcmean". As far as I understand "em" is the mean across all ensembles while "fcmean" is the value per ensemble 
+
+Then there also is the "numberOfPoints". This has to do with the geographical area, since we receive the forecast for several geographical areas. 
+Just by testing and looking at the latitude/longitude range I figured 384 is the one matching Malawi but no idea if there is a better way to set this
+
+```python
+xr.load_dataset(filepath_list[0],engine="cfgrib",filter_by_keys={'numberOfPoints': 384, 'dataType': 'fcmean'})
+```
+
+Next we load all files at once
+
+```python
+def _preprocess_monthly_mean_dataset(ds_date: xr.Dataset):
+        
+         return (
+            ds_date.assign_coords(
+                {
+                    #get the difference between the valid time month and the publication month 
+                    #valid time is one month ahead of the start of the actual valid time
+                    #so a leadtime of 1 means that the forecast is forecasting the same month as the publication month
+                    "step": (ds_date.valid_time.dt.month-ds_date.time.dt.month)+12*(ds_date.valid_time.dt.year-ds_date.time.dt.year)
```

Comment:
Here can't you do `(ds_date.valid_time.dt.month-ds_date.time.dt.month -1) % 12` ?

Reply:
Oops nevermind, the mod doesn't work and also as you explained, both `time` and `valid_time` are ahead by one month

Reply:
Due to the github issue solution we found this is no longer relevant. Changed the code to load the times correctly (and not use `valid_time` anymore)